test/Conversion/HandshakeToHW/test_simple_loop.mlir

```diff
@@ -0,0 +1,73 @@
+// RUN: circt-opt -lower-handshake-to-hw -split-input-file %s | FileCheck %s
+
+// CHECK-LABEL:   hw.module.extern @arith_addi_in_ui64_ui64_out_ui64(
+// CHECK-SAME:                                                       %[[VAL_0:.*]]: !esi.channel<index>,
+// CHECK-SAME:                                                       %[[VAL_1:.*]]: !esi.channel<index>) -> (out0: !esi.channel<index>)
+// CHECK:         hw.module.extern @handshake_cond_br_in_ui1_2ins_2outs_ctrl(%[[VAL_2:.*]]: !esi.channel<i1>, %[[VAL_3:.*]]: !esi.channel<none>) -> (outTrue: !esi.channel<none>, outFalse: !esi.channel<none>)
+// CHECK:         hw.module.extern @handshake_sink_in_ui64(%[[VAL_0]]: !esi.channel<index>)
+// CHECK:         hw.module.extern @handshake_cond_br_in_ui1_ui64_out_ui64_ui64(%[[VAL_2]]: !esi.channel<i1>, %[[VAL_3]]: !esi.channel<index>) -> (outTrue: !esi.channel<index>, outFalse: !esi.channel<index>)
```

Comment:
Seems like it would be nice to generate most of the specializations here automatically.  I guess as we add 'generic' support for these modules, they would become generated?  It's unclear to me whether it will be easier to specialize before lowering to hw.module, or to generate something generic here and then specialize all the instances here?   Eventually we'd probably want a scheme where we'd generate something generic, but perhaps override the generic implementation with something more specialized for certain cases.

Reply:
If there was a clear mechanism to perform generic/template lowering, i'd gladly use it - but as far as i am aware of, there is none (correct?). It could be interesting to add support for this in the hw.module operation; no matter what it's probably worth investigating before delving deeper into this commit series.

Reply:
This sort of reminds me of how the linalg dialect basically just has `linalg.generic`, but there are named ops you can add on top of that, like certain flavors of convolution. To handle things generically, would this lower to `hw.module` using `comb` and `seq`? It seems like we could do something similar to linalg where you could bring your own library of named ops, or fall back to generic `hw.modules`.